In [1]:
# Jupyter Notebook for baseline model training
import sys
sys.path.append('..')

import pandas as pd
import numpy as np

from src.data_loader import load_and_validate
from src.features import add_time_of_day_feature, train_test_split_stratified

# Reusing Day 1's validated loader — if the data is malformed, we find
# out here, at cell 1, not mid-training.
df = load_and_validate()
df = add_time_of_day_feature(df)
df.shape

(284807, 32)

In [2]:
# Splitting data in train and test sets, stratified by the target variable
X_train, X_test, y_train, y_test = train_test_split_stratified(df)

print(f"Train: {len(X_train)} rows, {y_train.sum()} fraud ({y_train.mean():.4%})")
print(f"Test:  {len(X_test)} rows, {y_test.sum()} fraud ({y_test.mean():.4%})")

Train: 227845 rows, 394 fraud (0.1729%)
Test:  56962 rows, 98 fraud (0.1720%)


## Model Comparison Setup

This notebook trains and compares three approaches to fraud detection:

1. **Logistic Regression** — simple, fast, interpretable supervised baseline
2. **XGBoost** — stronger supervised model, likely the best performer on offline metrics
3. **Isolation Forest** — unsupervised anomaly detection, trained only on non-fraud transactions

All three are evaluated on precision, recall, and PR-AUC — not accuracy, which is meaningless at a 0.17% fraud rate (see `01_eda.ipynb`).